### PR TITLE
Feat: user dynamic_data sync and import

### LINKED ISSUE #2246: [FEATURE] Sync dynamic data to server
**What?**
Dynamic data list overrides are currently only stored on the device. This data should be synced to the server as well. 

**Why?**
When user's dynamic data is available on the server it can be used for data analysis and account recovery (see #2245).

### PR DESCRIPTION
**Pull request by chrismclarke**
PR Checklist

- [x] PR title descriptive (can be used in release notes)

## Description
**Main Changes**
- Add debug page to view current user contact fields and dynamic data
- Add migration to include `dynamic_data` column with server db
- Update app sync and import methods to include dynamic_data

**Additional Changes**
- Add a debounce to reduce the frequency user dynamic data is written to disk (max once per 500ms). This seemed a sensible optimisation given that import operations happen 1 row and a time, so could easily trigger a significant number of operations which may block the main thread and make the app hang
- Remove API `/tables` endpoint (as suggested in #2277)

**NOTE**
Production changes will only take effect once merged and server code updated (follow-up task)

## Review Notes
In order to populate dynamic data recommend viewing http://localhost:4200/template/comp_data_items

User debug page available at  http://localhost:4200/user

Sync testing requires a running development server environment (API + DB)

## Dev Notes
We don't currently have a set of spec tests to add db persist debounce test to, and feels beyond scope of this PR to create all from scratch. 

The persist operation has been tested locally by adding console logs to `src\app\shared\services\dynamic-data\adapters\persistedMemory.ts` public `persistStateToDB` and private `handleStatePersist` methods, and checking when importing user data that the the `handleStatePersist` method is only invoked once despite multiple calls to `persistStateToDB`.

## Author Notes
The `/user` debug page hasn't been included in the debug content app menu (but could in the future)

The list of dynamic_data is formatted as a single column in the existing `app_users` table, nested first by flow_type (for now all are data_lists) and then by flow_name. E.g. for the synced user the `dynamic_data` column entry is formatted as:
```json
{
  "data_list": {
    "comp_data_items_list": {
      "id_1": {
        "id": "id_1",
        "label": "Task 1",
        "completed": true
      },
      "id_2": {
        "id": "id_2",
        "label": "Task 2",
        "completed": false
      },
      "id_3": {
        "id": "id_3",
        "label": "Task 3",
        "completed": false
      }
    }
  }
}
```

## Git Issues

Closes #2246

## Screenshots/Videos
**User Sync and Import demo**
Create dynamic data entries using the `comp_data_items` template, and view available user data on new user debug page. 
Sync user data to db (running locally), and then clear cache to generate a new user profile.
Import initial user to see contact fields and dynamic data restored.

[Screenity video - Apr 19, 2024.webm](https://github.com/IDEMSInternational/parenting-app-ui/assets/10515065/ae05c643-de78-438c-9b38-d91201dc7896)

**Example DB**
User entry now includes `dynamic_data` column with user data
image

### CODE REVIEW

**Review by esmeetewinkel — COMMENTED** [review]
I first tried testing locally, but struggled to get my local API set up. I then tested on the web deployment preview (but I'm not sure whether it's expected to work there!) 

The `/user` page seems to work correctly (I could import a user succesfully), but importing the dynamic data did not work. I also couldn't find the dynamic data on the server database. What table should this sit in? 

N.B. I was seeing a console error 
```
error-handler.service.ts:65 ERROR TypeError: Cannot read properties of undefined (reading 'persistStateToDB')
```
[Screenity video - Apr 26, 2024.webm](https://github.com/IDEMSInternational/parenting-app-ui/assets/74557272/18ebca72-fbf0-4288-add1-f92e8363c148)

**Comment by chrismclarke** [comment]
> I first tried testing locally, but struggled to get my local API set up. I then tested on the web deployment preview (but I'm not sure whether it's expected to work there!)

> The `/user` page seems to work correctly (I could import a user succesfully), but importing the dynamic data did not work. I also couldn't find the dynamic data on the server database. What table should this sit in?

Unfortunately it won't work on web deployment until I've deployed the changes there, so for now the only way to test is to have a database running locally. When running locally (or after future deployment) the data will sit in the same `app_users` table as the contact_fields

> N.B. I was seeing a console error
> 
> ```
> error-handler.service.ts:65 ERROR TypeError: Cannot read properties of undefined (reading 'persistStateToDB')
> ```

That's weird, was this on the user page or somewhere else? Do you know if it appeared on first load or after you clicked to sync/import?

File: src/app/feature/user/pages/user-debug/user-debug.page.html [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        <tr>
+          @for(header of table.headers; track header){
+          <th>{{header}}</th>
+          }
+        </tr>
+      </thead>
+      <tbody>
+        @for(row of table.rows; track row.id){
+        <tr>
+          @for(header of table.headers; track header){
```
Comment (jfmcquade) [inline src/app/feature/user/pages/user-debug/user-debug.page.html:54]: The Angular 17 syntax looking nice and readable

File: src/app/feature/user/pages/user-debug/user-debug.page.html [hunk 2]
```diff
@@ -0,0 +1,64 @@
+<ion-content class="ion-padding">
+  <div style="display: flex; align-items: center; gap: 8px">
+    <span>App User ID:</span>
+    <ion-input [(ngModel)]="importUserId" style="flex: 1"></ion-input>
+    <ion-button (click)="importUserData()" [disabled]="importUserId===userId">Import</ion-button>
```
Comment (jfmcquade) [inline src/app/feature/user/pages/user-debug/user-debug.page.html:5]: I'm not sure we necessarily want to disable when the IDs match – we might want to overwrite the local user data with that from when the same user last synced? I'm not sure what the real world use case would be, but as this is a debug page that might be worth allowing as it's a possible use of the `user: import` action (that's how I tested the functionality on #2270, see [comment](https://github.com/IDEMSInternational/parenting-app-ui/pull/2270#pullrequestreview-1980363795)).

**Review by jfmcquade — COMMENTED** [review]
Looks good, I can confirm that sync and export and export the user data including the dynamic data, and it displays in the app as expected.

Screenshot 2024-05-02 at 11 40 14

Screenshot 2024-05-02 at 11 39 14

One minor comment inline.

Also, I get the same error that Esmee documented above when viewing the hosted PR preview, on the initial load of the app home page. I haven't seen it locally though. Does the `writeCache` (`PersistedMemoryAdapter`) need some kind of initialisation method that can be awaited by the dynamic data service before calling its methods? Or else perhaps the services that call `dynamicDataService.getState()` (user meta and server) should be awaiting `dynamicDataService.ready()` before doing so?

Screenshot 2024-05-02 at 12 16 03

**Comment by chrismclarke** [comment]
> Also, I get the same error that Esmee documented above when viewing the hosted PR preview, on the initial load of the app home page. I haven't seen it locally though. Does the `writeCache` (`PersistedMemoryAdapter`) need some kind of initialisation method that can be awaited by the dynamic data service before calling its methods? Or else perhaps the services that call `dynamicDataService.getState()` (user meta and server) should be awaiting `dynamicDataService.ready()` before doing so?

Most likely, I've just gone through all the services that import the dynamicDataService and added 33ab0d4 to try address any potential issues - I think error should be resolved now (not seeing the console error on preview link)

File: src/app/feature/user/pages/user-debug/user-debug.page.html (hunk 2 above)
Comment (chrismclarke) [reply src/app/feature/user/pages/user-debug/user-debug.page.html]: Good idea, resolved in d8fb88fd627db8f7255fa0165104b5df70a6c47f

**Review by jfmcquade — APPROVED** [review]
> Most likely, I've just gone through all the services that import the dynamicDataService and added [33ab0d4](https://github.com/IDEMSInternational/parenting-app-ui/commit/33ab0d4f72565cc24ad5dc49aac5eb3578c6e440) to try address any potential issues - I think error should be resolved now (not seeing the console error on preview link) @esmeetewinkel - let me know if you still see

Looks good, I can confirm that the error is no longer thrown when I visit the PR web preview link. I think this should be good to merge

**Review by esmeetewinkel — COMMENTED** [review]
I'm still seeing the error on the web preview (which I checked has been updated for the last commit) when loading the `/user` page 

image

**Comment by chrismclarke** [comment]
> I'm still seeing the error on the web preview (which I checked has been updated for the last commit) when loading the `/user` page
> 
> ![image](https://private-user-images.githubusercontent.com/74557272/327718459-299b8743-aa37-436a-99ce-

Thanks @esmeetewinkel - I had only been checking the home page, not the user page but now I can see the same error. It related to the server sync service which didn't properly wait for the dynamic data service to be ready before trying to read the user tables from it. Fixed with aedfbcd

**Review by esmeetewinkel — APPROVED** [review]
Can confirm the bug on the `/user` page disappeared

**Pull request merged**